### PR TITLE
feat(Widgets): FadeBin

### PR DESCRIPTION
**Pull request by GeopJr**
One of these days I'll make libgeopjr or something because it's a shame for all these custom widgets to not be used in other apps.

Anyway, faded view. Not sure if we should go forward with this but it's configurable enough to be able to auto-reveal with a setting or similar.

https://github.com/user-attachments/assets/19a02fde-db3e-40a5-b35c-25d2c8ef8d32

It could also come with a configurable max-height but for now it's a constant.

### CODE REVIEW

**Comment by GeopJr** [comment]
@bugaevc quick question, would this have any impact on hitting the fast path?

FadeBin is HEIGHT_FOR_WIDTH unless there's a child, which then uses the child's mode. In statuses, there's always a child, the markupview.

The rest is mainly setting the min (<=) and nat (vertical) sizes to 300. Allocation passes to the child.

File: src/Widgets/FadeBin.vala [hunk 1]
```diff
@@ -0,0 +1,150 @@
+public class Tuba.Widgets.FadeBin : Gtk.Widget {
+	const int MAX_HEIGHT = 300;
+	const float FADE_HEIGHT = 125f;
+
+	private Gtk.Widget? _child = null;
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:5]: `unowned`

File: src/Widgets/FadeBin.vala [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	public Gtk.Widget? child {
+		get { return _child; }
+		set {
+			if (_child != null) _child.unparent ();
+			_child = value;
+			_child.set_parent (this);
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:11]: null check?

File: src/Widgets/FadeBin.vala [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		}
+	}
+
+	~FadeBin () {
+		if (this.child != null) this.child.unparent ();
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:16]: This should be done in [`dispose`](https://valadoc.org/gobject-2.0/GLib.Object.dispose.html), something like this:

```vala
public override void dispose () {
    this.child = null;
    base.dispose ();
}
```

File: src/Widgets/FadeBin.vala [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	public override Gtk.SizeRequestMode get_request_mode () {
+		if (this.child != null) return this.child.get_request_mode ();
+		return Gtk.SizeRequestMode.HEIGHT_FOR_WIDTH;
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:21]: If you don't have a child, return `CONSTANT_SIZE`.

File: src/Widgets/FadeBin.vala [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	private bool _reveal = false;
+	public bool reveal {
+		get { return _reveal; }
+		set {
+			if (_reveal != value) {
+				_reveal = value;
+				this.queue_draw ();
+				this.queue_resize ();
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:31]: `queue_resize ()` implies `queue_draw ()`, no point in calling both

File: src/Widgets/FadeBin.vala [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	const Gsk.ColorStop[] GRADIENT = {
+		{ 0f, { 0, 0, 0, 1f } },
+		{ 1f, { 0, 0, 0, 0f } },
+	};
+
+	public override void size_allocate (int width, int height, int baseline) {
+		if (this.child == null) {
+			base.size_allocate (width, height, baseline);
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:60]: The default implementation does nothing, so normally nobody bothers to call it.

File: src/Widgets/FadeBin.vala [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+			return;
+		}
+
+		this.child.allocate (width, height, baseline, null);
+		if (this.reveal) {
+			this.should_fade = false;
+			return;
+		}
+
+		this.should_fade = this.child.get_height () >= MAX_HEIGHT;
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:71]: Don't use `get_height ()` like that, it's for the widget itself to call. Use `get_allocated_height ()` (which is deprecated I think?), or better, `height`, since that's what you've just allocated.

EDIT: `child_height` not `height`, if you follow the suggestion above.

File: src/Widgets/FadeBin.vala [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+	}
+
+	public override void measure (Gtk.Orientation orientation, int for_size, out int minimum, out int natural, out int minimum_baseline, out int natural_baseline) {
+		this.child.measure (
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:75]: null check?

File: src/Widgets/FadeBin.vala [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+			out minimum_baseline,
+			out natural_baseline
+		);
+		if (this.reveal) return;
+
+		minimum = int.min (minimum, MAX_HEIGHT);
+		minimum_baseline = -1;
+		natural_baseline = -1;
+		if (orientation == Gtk.Orientation.VERTICAL) {
+			natural = MAX_HEIGHT;
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:89]: `int.min (natural, MAX_HEIGHT)` rather

File: src/Widgets/FadeBin.vala [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+			orientation,
+			for_size,
+			out minimum,
+			out natural,
+			out minimum_baseline,
+			out natural_baseline
+		);
+		if (this.reveal) return;
+
+		minimum = int.min (minimum, MAX_HEIGHT);
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:85]: Only for `VERTICAL`, so move this into the `if` below.

File: src/Widgets/FadeBin.vala [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+		minimum_baseline = -1;
+		natural_baseline = -1;
+		if (orientation == Gtk.Orientation.VERTICAL) {
+			natural = MAX_HEIGHT;
+		}
+	}
+
+	public override void snapshot (Gtk.Snapshot snapshot) {
+		if (this.child == null || !this.faded) {
+			this.child.snapshot (snapshot);
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala:95]: Again this looks wrong wrt null.

Also: never use `child.snapshot (snapshot)`; it's wrong of Vala to even expose it (file a bug, perhaps). Always use `snapshot_child (child, snapshot)`.

**Comment by GeopJr** [comment]
Thanks for the reviews! If you conclude that this is not feasible in a performant way, I'm okay with dropping it.

File: src/Widgets/FadeBin.vala (hunk 1 above)
Comment (GeopJr) [reply src/Widgets/FadeBin.vala:5]: Honestly, I'm scared of using unowned / weak ([`At the moment weak and unowned can be used interchangeably`](https://wiki.gnome.org/Projects/Vala/ReferenceHandling)) anymore. Every segfault Tuba has had is due to using them because, unless there's some secret way to do this that I don't know of, null checking is not enough and will happily pass them if it has been freed and then segfault.

Latest #1340 #1347 (also, both cases are very weird since the status should still hold a reference to them anyway!).

I guess it won't hurt testing it out and see if we hit it.

File: src/Widgets/FadeBin.vala [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+			return;
+		}
+
+		Gtk.Snapshot child_snapshot = new Gtk.Snapshot ();
+		this.snapshot_child (this.child, child_snapshot);
+		var node = child_snapshot.to_node ();
+		if (node == null) {
+			this.child.snapshot (snapshot);
+			return;
+		}
```
[the remark this answers is not in the document]
Comment (GeopJr) [reply src/Widgets/FadeBin.vala:111]: The quoted snippet is for when there are no nodes. The snapshot_child before it, is for getting the node bounds. I can change it to free_to_node as we have no use for it afterwards, but we do need the node bounds for gradient.

**Comment by bugaevc** [comment]
I do think it is both feasible and interesting, please keep going :slightly_smiling_face:

File: src/Widgets/FadeBin.vala (hunk 12 above)
[the remark this answers is not in the document]
Comment (bugaevc) [reply src/Widgets/FadeBin.vala:124]: You shouldn't need the node bounds either, something like this should work:

```vala
var width = get_width ();
var height = get_height ();

snapshot.push_mask (Gsk.MaskMode.INVERTED_ALPHA);
snapshot.append_linear_gradient (
  { { 0, height - FADE_HEIGHT }, { width, FADE_HEIGHT } },
  { 0, height },
  { 0, height - FADE_HEIGHT },
  GRADIENT
);

snapshot.pop ();
snapshot_child (this.child, snapshot);
snapshot.pop ();
```

[the remark this answers is not in the document]
Comment (GeopJr) [reply src/Widgets/FadeBin.vala:124]: Nope :(

image

File: src/Widgets/FadeBin.vala [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+		this.should_fade = this.child.get_height () >= MAX_HEIGHT;
+	}
+
+	public override void measure (Gtk.Orientation orientation, int for_size, out int minimum, out int natural, out int minimum_baseline, out int natural_baseline) {
+		this.child.measure (
+			orientation,
+			for_size,
```
[the remark this answers is not in the document]
Comment (GeopJr) [reply src/Widgets/FadeBin.vala]: Thanks :heart:

File: src/Widgets/FadeBin.vala (hunk 12 above)
[the remark this answers is not in the document]
Comment (bugaevc) [reply src/Widgets/FadeBin.vala:128]: Ah well, you should clip to your own bounds then. `{ 0, 0, width, height }` for starters, but you're probably only going to want to clip to the lower edge to keep things like shadows working otherwise.

[the remark this answers is not in the document]
Comment (GeopJr) [reply src/Widgets/FadeBin.vala:128]: I mean, give me a second to play around with clipping to the same rectangle.

File: src/Widgets/FadeBin.vala [hunk 14]
```diff
@@ -80,21 +80,25 @@ public class Tuba.Widgets.FadeBin : Gtk.Widget {
...
+			child_for_size,
 			out minimum,
 			out natural,
 			out minimum_baseline,
 			out natural_baseline
 		);
-		if (this.reveal) return;
 
-		minimum_baseline = -1;
-		natural_baseline = -1;
```
Comment (bugaevc) [inline src/Widgets/FadeBin.vala]: I meant, keep setting these to `-1` for `!reveal` branch. Not that it matters.

**Comment by GeopJr** [comment]
> I'd expect the fade/size to animate when expanded, like Gtk.Revealer does. Do you not want that?

It might be too much for very long posts but I'll see about it. If we go for it, I'd probably guess:
- snapshot whole child no fade
- overflow hidden
- set min/nat to `int.max (minimum * animation.value, MAX_HEIGHT);`

**Comment by bugaevc** [comment]
> set min/nat to `int.max (minimum * animation.value, MAX_HEIGHT);`

It's a lot more complex than that. Basically, you want to animate the height limit from `MAX_HEIGHT` up to "child's actual height". Again, I would have to think for (a much longer) while about the correct way here. It's probably best if you don't block this PR on the animation, we can add it later.

**Comment by GeopJr** [comment]
> we can add it later

:+1: 

This PR is not done anyway, it was an initial implementation. I still need to add settings and polish the UI, strings etc.

**Comment by bugaevc** [comment]
Here's a working sketch of the animation
https://gitlab.gnome.org/-/snippets/6864

**Comment by GeopJr** [comment]
Thanks! Is there a specific reason on doing the animation manually instead of using AdwTimedAnimation?

**Comment by bugaevc** [comment]
Ah, sure, feel free to use that. I was focusing on the implementation of measure & size allocate, not on everything else.

**Comment by GeopJr** [comment]
I don't think I skipped anything, but the main difference is that FadeBin's API needs a way to expand without animation for cases like users having collapsing long posts off and pre-expanding them when opening a thread, so there's the `reveal` property and the `reveal_animated` (and `hide_animated`) functions. When the animation ends, the reveal property inverts.

https://github.com/user-attachments/assets/3637f854-7be7-4d52-99e5-3b46ffcadb76

**Comment by bugaevc** [comment]
* Does it work — does it animate nicely? Can you record a video? :smiley: Ideally, I should build this branch, slow down the animation, and verify that nothing about the layout invariants breaks.
* Not sure if the `LOTS` thing causes GSK to render a huge offscreen texture, or if it already does optimizations based on bounds etc.

**Comment by bugaevc** [comment]
I see, cool :smile:

**Comment by GeopJr** [comment]
> Not sure if the LOTS thing causes GSK to render a huge offscreen texture, or if it already does optimizations based on bounds etc.

I wonder if extreme edge cases will have any impact (100k+ character long posts) :thinking: Will test it out later.

Thanks again for everything! :bow:
